### PR TITLE
1 psr cache

### PR DESCRIPTION
**Pull request by spotman**
Basic PSR-16 implementation done. Couple of methods was deprecated coz there is no direct access to cache file anymore. Basic logger was extracted to DefaultLogger with PSR-3 interface. Also "headers forwarding" feature implemented so headers from application response can be stored into cache item now. There is no tests yet, please make breif review for these changes. Thanks!

### CODE REVIEW

**Comment by mmamedov** [comment]
Hi @spotman, cool changes thanks! But it is too much of code changes, and I will need more time to digest it. My suggestion is to make it in small steps. In that way it will be easier to test and safe to merge. 

**Let's divide it into small steps.** How about you add new classes for PSR-16, and loggers, without changing any existing code? Push a new branch (or edit current, as you wish) with only these changes, and I will merge it under "**v2.0-dev**" branch

**So it would be the following classes - CachItem/Interface, FileSystemCacheAdapter, DefaultLogger, Exception classes, and composer.json.** 

On the top of each new file please add comments following the current template, with your name (it should be first line, before namespaces). No need to change years on files that were created last year, let it stay for visibility.

Once you do this I can merge it fast, and we can continue refactoring. 
I like where we are going with this, I think it will be great!

File: README.md [hunk 1]
```diff
@@ -101,7 +101,7 @@ The following are public methods of PageCache class that you could call from you
 - disableSession():void - Disable session support.
 - sessionExclude(array):void - Exclude $_SESSION key(s) from caching strategies 
 - isCached():bool - Checks if current page is in cache, true if exists false if not cached yet.
-- getFilePath():string - Get full path for current page's filename. At this point file itself might or might not have been created.
+- getFilepath():string - Get full path for current page's filename. At this point file itself might or might not have been created.
```
Comment (mmamedov) [inline README.md]: Let's not change this. Filepath are two separate words, let's keep it getFilePath

File: examples/demo-api.php [hunk 2]
```diff
@@ -30,7 +30,7 @@
 $cache->enableLog();
 
 //getfilepath
-echo 'Cache filepath, getFilePath(): ' . $cache->getFilePath() . '<br/>';
+echo 'Cache filepath, getFilepath(): ' . $cache->getFilePath() . '<br/>';
```
Comment (mmamedov) [inline examples/demo-api.php]: idem

File: src/Storage/FileSystem/FileSystem.php [hunk 3]
```diff
@@ -1,4 +1,6 @@
 <?php
+namespace PageCache\Storage\FileSystem;
```
Comment (mmamedov) [inline src/Storage/FileSystem/FileSystem.php]: This should have stayed where it was, after comments block

File: src/Storage/FileSystem/FileSystem.php [hunk 4]
```diff
@@ -105,7 +107,7 @@ public function writeAttempt()
         /**
          * File locking disabled?
          */
-        if (empty($this->file_lock)) {
+        if ($this->file_lock === null) {
```
Comment (mmamedov) [inline src/Storage/FileSystem/FileSystem.php:110]: Do we need to do this? empty() covers it already.

File: src/Storage/FileSystem/FileSystem.php [hunk 5]
```diff
@@ -160,7 +162,7 @@ public function writeAttempt()
     public function setFileLock($file_lock)
     {
         if (empty($file_lock)) {
-            throw new \Exception(__CLASS__ . ' file lock can not be empty');
+            throw new PageCacheException(__CLASS__ . ' file lock can not be empty');
```
Comment (mmamedov) [inline src/Storage/FileSystem/FileSystem.php:165]: Nice job with PageCacheException!

File: src/Storage/FileSystem/HashDirectory.php [hunk 6]
```diff
@@ -1,4 +1,5 @@
 <?php
+namespace PageCache\Storage\FileSystem;
```
Comment (mmamedov) [inline src/Storage/FileSystem/HashDirectory.php]: Let's keep it as it was, after comments.

File: src/Storage/FileSystem/HashDirectory.php [hunk 7]
```diff
@@ -56,9 +57,9 @@ public function __construct($file = null, $dir = null)
      */
     public function setDir($dir)
     {
-        if (!empty($dir)) {
+        if ($dir) {
```
Comment (mmamedov) [inline src/Storage/FileSystem/HashDirectory.php:60]: !empty() is prefered here, make it more readable.

File: src/Strategy/MobileStrategy.php [hunk 8]
```diff
@@ -40,10 +40,10 @@ class MobileStrategy implements StrategyInterface
      */
     public function __construct(\Mobile_Detect $mobileDetect = null)
     {
-        if (is_null($mobileDetect)) {
-            $this->MobileDetect = new \Mobile_Detect;
-        } else {
+        if ($mobileDetect !== null) {
```
Comment (mmamedov) [inline src/Strategy/MobileStrategy.php]: During refactoring we can replace whole if block with:
`$this->MobileDetect = $mobileDetect ?: new \Mobile_Detect;
`

**Review by mmamedov — CHANGES_REQUESTED** [review]
As we discussed, let's split these changes into smaller chunks.

File: src/Storage/FileSystem/FileSystem.php (hunk 4 above)
Comment (spotman) [reply src/Storage/FileSystem/FileSystem.php:110]: I use "Php Inspections (EA extended)" plugin for PhpStorm, it recommends to replace "empty" with null check. https://github.com/kalessil/phpinspectionsea/blob/master/docs/types-compatibility.md#empty-usage Also I've read some years ago that null check is much faster. I may change it back, up to you.

File: README.md (hunk 1 above)
Comment (spotman) [reply README.md]: Hmm, it's refactoring bug probably coz PageCache has method `getFilePath` with capital `P`. I'll change this line in readme.

**Comment by spotman** [comment]
Hi @mmamedov!

PSR cache implementation would affect all of these files except `HttpHeader(s)` and `DefaultLogger`. Also methods `PageCache::getFile()` and `PageCache::getFilePath()` would be deprecated cos no direct file access would be allowed via PSR-16 cache interface. So I think that splitting would not have any valuable effect coz implementing PSR interfaces needs a huge changes already. I may extract feature for headers forwarding and Expires header but it is also relatively small change. As said in PM, I've tested all of these changes "by hands" and everything worked. I agree with you that it needs more testing and unittests may be written for these new cases. There is an [integration tests suite for PSR-16 providers](https://github.com/matthiasmullie/psr16-integration-tests)  so we may use it for testing `FileSystemCacheAdapter` and underlying classes like `PageCache\Storage\FileSystem\FileSystem`. Any other classes still need to be tested and I may write tests for them if you agree with overall class structure.

**Pull request merged**